src/common.h

```diff
@@ -23,10 +26,61 @@
 #endif
 #endif
 
-#define ARRAYS_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
 
 #define MASK(n) (~((~0U << (n))))
 
+#if defined(_MSC_VER)
+#include <intrin.h>
+static inline int rv_clz(uint32_t v)
+{
+    /* 0 is considered as undefined behavior */
+    assert(v);
+
+    uint32_t leading_zero = 0;
+    _BitScanReverse(&leading_zero, v);
+    return 31 - leading_zero;
+}
+#elif defined(__GNUC__) || defined(__clang__)
+static inline int rv_clz(uint32_t v)
+{
+    /* https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html */
+    /* 0 is considered as undefined behavior */
+    assert(v);
+
+    return __builtin_clz(v);
+}
+#else /* generic implementation */
+static inline int rv_clz(uint32_t v)
+{
+    /* http://graphics.stanford.edu/~seander/bithacks.html#IntegerLogDeBruijn */
+    static const uint8_t mul_debruijn[] = {
+        0, 9,  1,  10, 13, 21, 2,  29, 11, 14, 16, 18, 22, 25, 3, 30,
+        8, 12, 20, 28, 15, 17, 24, 7,  19, 27, 23, 6,  26, 5,  4, 31,
+    };
+
```

Comment:
Since msc, gcc, and clang implementations include `assert(v)`, let's also add `assert(v)` in the fallback implementation?